### ISSUE TITLE
[Feature Request] Hubic Support

### ISSUE THREAD

**Issue by EricThi** [keeweb/keeweb#147]
Hello,

It's possible to sync kbx with hubic?

hubic has API : https://hubic.com/home/browser/developers/

Eric

**Comment by antelle** [comment]
Support can be added later but as it's not popular - I think, only as an external plugin. For now there will be GDrive, OneDrive and WebDAV.

**Comment by lucifer2791** [comment]
mmm, I don't use another than Hubic because of PatrioctAct...

At worst, I will wait WebDAV for me usage (or stock on local to my server )

Thank you for your answerd and Have Fun ;)

Eric

**Comment by antelle** [comment]
Just wondering: why Hubic is preferred because of Patriot Act? There're no compliance notes and documents on their website, they don't offer self-hosting plans, no audits were performed, it's not u.s. company. Sorry for stupid questions, it's the first time I'm told about this storage.

**Comment by lucifer2791** [comment]
Hubic = OVH
Hubic it's only in France = no patrioct Act

Eric

**Comment by antelle** [comment]
Oh, you mean that patriot act requires some surveillance/monitoring requirements and other providers are not so secure, right?

**Comment by lucifer2791** [comment]
They do not hide.
- http://kairos-creations.fr/cloud-match-google-drive-onedrive-en-4-points/ (fr)
- http://www.zdnet.com/article/google-admits-patriot-act-requests-handed-over-european-data-to-u-s-authorities/ (en)

All data on U.S. and some countries on EU ( not in France :D)

Eric

**Comment by kasoban** [comment]
His argument seems to be that Hubic as a France based french company is not bound to the Patriot Act, and thus is not expected to hand over his data to US authorities on request.

But since the latest Paris attacks France is on the best way to equally draconian privacy laws, so I wouldn't keep my hopes up in that regard anyways...

**Comment by grinsted** [comment]
I am no expert, but as i understand it then hubiC uses a variant of the OpenStack Swift API. I believe that is why hubiC is in so many other open source projects such as duplicati and rclone. So i guess that hubiC support would be almost there if S3 support is implemented.

**Comment by nnjeremy** [comment]
Hi,
I use hubic everyday and I hope we can add this storage to keeweb.
If it's needed I can help for the external plugin.

**Comment by antelle** [comment]
@nnjeremy ok, thanks, once the storage providers architecture is ready with a working example of GDrive I'll write here.

**Comment by lucifer2791** [comment]
@grinsted : yes of course : https://www.ovh.com/fr/a1134.pour_hubic_ovhcom_adapte_openstack (fr) or https://github.com/oderwat/hubic2swiftgate

@kasoban : patrioct act = data on U.S.A.
Hubic it's only on France = French legislation (OVH post (fr) :
- https://www.ovh.com/fr/a1743.le-gouvernement-veut-il-contraindre-les-hebergeurs-internet-a-l-exil
- https://www.ovh.com/fr/a1766.point-vue-ovh-loi-renseignement
another (fr)
http://www.journaldugeek.com/2015/05/12/ovh-reagit-loi-renseignement-pas-de-boites-chez-hebergeurs/

In France, for scan data on DC : need request to procurator + mail and/or ip on server :)

@nnjeremy & @antelle : thank you for accept my request :+1:  \Ô/

**Comment by antelle** [comment]
@nnjeremy  storage providers have been implemented: [google drive](https://github.com/antelle/keeweb/blob/develop/app/scripts/storage/storage-gdrive.js), [onedrive](https://github.com/antelle/keeweb/blob/develop/app/scripts/storage/storage-onedrive.js). You can create something like this for hubic. Please:
1. don't add any dependencies
2. make it disabled by default
3. use existing API for xhr and OAuth

**Comment by nnjeremy** [comment]
@antelle thanks for your work ! I will look at it next week.

**Comment by Servuc** [comment]
Hi,

Somebody has work on that ? 
I'm interesting to work on it, for the same reason, no patriot act in France ! (Only for DC)

If nobody work on it, go fork and PR :smile: 

Have a good coding day !

**Comment by antelle** [comment]
@Servuc, I guess, no. I'm not planning to work on it, and it looks like, nobody is working on it, as @nnjeremy's comment was quite long ago.

**Comment by Servuc** [comment]
Thanks ;) 
I try it, and gave you more information in one week !

**Comment by antelle** [comment]
Ok, please refer to my [previous comment](https://github.com/keeweb/keeweb/issues/147#issuecomment-202089757) about the implementation details.

**Comment by Servuc** [comment]
Hi,

I've begin slowly (I does that during my eating time at work :smile: ), 

First, I've done icon ! 

![capture du 2016-07-28 14 17 56](https://cloud.githubusercontent.com/assets/10000785/17212195/1d0e4fc0-54ce-11e6-99be-0ee66babd164.png)

(It's a screenshot for GitHub) 

Have a good code !

**Comment by antelle** [comment]
Now it's possible to create custom storage as a plugin.
Example: https://github.com/keeweb/keeweb/tree/develop/plugins/examples/fail-storage

**Comment by antelle** [comment]
Closing. Feel free to submit a plugin, no work will be done in KeeWeb.

**Comment by antelle** [comment]
Added a label, to track plugin requests. Let's leave it open...
